pkg/policy/rule.go

```diff
@@ -35,54 +35,93 @@ func (r *rule) String() string {
-		log.WithFields(logrus.Fields{
-			logfields.EndpointSelector: endpoints,
-			"policy":                   policy,
-		}).Debug("skipping L4 filter as the endpoints are already covered.")
-		return true
-	}
-
-	if len(policy.Endpoints) > 0 && len(endpoints) == 0 {
-		log.WithFields(logrus.Fields{
-			logfields.EndpointSelector: endpoints,
-			"policy":                   policy,
-		}).Debug("new L4 filter applies to all endpoints, making the policy more permissive.")
-		policy.Endpoints = nil
-	}
-
-	policy.Endpoints = append(policy.Endpoints, endpoints...)
-	return false
-}
-
+// mergeL4IngressPort merges all rules which share the same port & protocol that
+// select a given set of endpoints. It updates the L4Filter mapped to by the specified
+// port and protocol with the contents of the provided PortRule. If the rule
+// being merged has conflicting L7 rules with those already in the provided
+// L4PolicyMap for the specified port-protocol tuple, it returns an error.
 func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r api.PortRule, p api.PortProtocol,
 	proto api.L4Proto, ruleLabels labels.LabelArray, resMap L4PolicyMap) (int, error) {
 
 	key := p.Port + "/" + string(proto)
-	filter, ok := resMap[key]
+	existingFilter, ok := resMap[key]
 	if !ok {
 		resMap[key] = CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
 		return 1, nil
 	}
-	l4Filter := CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
-	if l4Filter.L7Parser != "" {
-		if filter.L7Parser == "" {
-			filter.L7Parser = l4Filter.L7Parser
-		} else if l4Filter.L7Parser != filter.L7Parser {
-			ctx.PolicyTrace("   Merge conflict: mismatching parsers %s/%s\n", l4Filter.L7Parser, filter.L7Parser)
-			return 0, fmt.Errorf("Cannot merge conflicting L7 parsers (%s/%s)", l4Filter.L7Parser, filter.L7Parser)
+
+	// Create a new L4Filter based off of the arguments provided to this function
+	// for merging with the filter which is already in the policy map.
+	filterToMerge := CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
+
+	// Handle cases where filter we are merging new rule with, or new rule itself
+	// allows all traffic on L3.
+	//
+	// Existing filter selects all endpoints, so don't add new endpoints.
+	if len(existingFilter.Endpoints) == 0 && len(filterToMerge.Endpoints) > 0 {
+		log.WithFields(logrus.Fields{
+			logfields.EndpointSelector: filterToMerge.Endpoints,
+			"policy":                   existingFilter,
+		}).Debug("skipping L4 filter as the endpoints are already covered")
+
+		// Existing L4Filter already selects all endpoints; if there are no L7 rules
+		// to add from the new rule, then we can just exit because the rule is as
+		// permissive as possible at L3, the L4 information is already contained
+		// within the filter, and there is no L7 metadata to add to the filter.
+		if r.NumRules() == 0 {
+			existingFilter.DerivedFromRules = append(existingFilter.DerivedFromRules, ruleLabels)
+			return 1, nil
```

Comment:
Needs to write `existingFilter` back into `resMap` like in other places

Reply:
We actually don't technically have to here, because the filter already in the map selects all endpoints, and we have no L7-related metadata to update. But, I will do it for consistency's sake.

Reply:
Fixed in latest upload.